lib/SIL/IR/SILSymbolVisitor.cpp

```diff
@@ -0,0 +1,822 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements liker symbol enumeration for SILSymbolVisitor.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/SIL/SILSymbolVisitor.h"
+
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTMangler.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/FileUnit.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/PropertyWrappers.h"
+#include "swift/AST/SynthesizedFileUnit.h"
+#include "swift/SIL/FormalLinkage.h"
+#include "swift/SIL/SILLinkage.h"
+#include "swift/SIL/SILModule.h"
+#include "swift/SIL/SILVTableVisitor.h"
+#include "swift/SIL/SILWitnessTable.h"
+#include "swift/SIL/SILWitnessVisitor.h"
+
+using namespace swift;
+
+static bool requiresLinkerDirective(Decl *D) {
+  for (auto *attr : D->getAttrs()) {
+    if (auto *ODA = dyn_cast<OriginallyDefinedInAttr>(attr)) {
+      auto Active = ODA->isActivePlatform(D->getASTContext());
+      if (Active.hasValue())
+        return true;
+    }
+  }
+  return false;
+}
+
+static bool isGlobalOrStaticVar(VarDecl *VD) {
+  return VD->isStatic() || VD->getDeclContext()->isModuleScopeContext();
+}
+
+using DynamicKind = SILSymbolVisitor::DynamicKind;
+
+static Optional<DynamicKind> getDynamicKind(ValueDecl *VD) {
+  if (VD->shouldUseNativeMethodReplacement())
+    return DynamicKind::Replaceable;
+
+  if (VD->getDynamicallyReplacedDecl())
+    return DynamicKind::Replacement;
+
+  return None;
+}
+
+class SILSymbolVisitorImpl : public ASTVisitor<SILSymbolVisitorImpl> {
+  SILSymbolVisitor &Visitor;
+  const SILSymbolVisitorContext &Ctx;
+
+  /// A set of original function and derivative configuration pairs for which
+  /// derivative symbols have been emitted.
+  ///
+  /// Used to deduplicate derivative symbol emission for `@differentiable` and
+  /// `@derivative` attributes.
+  llvm::DenseSet<std::pair<AbstractFunctionDecl *, AutoDiffConfig>>
+      AddedDerivatives;
+
+  void addMainIfNecessary(FileUnit *file) {
+    // HACK: 'main' is a special symbol that's always emitted in SILGen if
+    //       the file has an entry point. Since it doesn't show up in the
+    //       module until SILGen, we need to explicitly add it here.
+    //
+    // Make sure to only add the main symbol for the module that we're emitting
+    // TBD for, and not for any statically linked libraries.
+    if (!file->hasEntryPoint() || file->getParentModule() != Ctx.getModule())
+      return;
+
+    auto entryPointSymbol =
+        file->getParentModule()->getASTContext().getEntryPointFunctionName();
+
+    if (auto *decl = file->getMainDecl()) {
+      addFunction(SILDeclRef::getMainDeclEntryPoint(decl),
+                  /*ignoreLinkage=*/true);
```

Comment:
The previous TBDGen implementation bypassed its own linkage/visibility checks for this symbol (though it did so subtly by providing an explicit mangling). If I didn't do this the existing tests broke. I suspect that the linkage computation used to filter out non-public symbols is insufficiently sophisticated. I'll add the same FIXME to the documentation for `addFunction()` in this file as I did for `addLinkEntity()` in `IRSymbolVisitor` to remind us that there's something wonky about needing this flag.

Reply:
Ah makes sense! A `SILDeclRef::Kind::EntryPoint` should always have `Public` linkage AFAIK, so I suspect this is because `SILDeclRef::getSubclassScope` needs to be taught to ignore entry point SILDeclRefs. Anyways, I'm happy to explicitly preserve the prior behavior for now.